tutorial/nixify/1_nix_flake.md

```diff
@@ -0,0 +1,223 @@
+
+#### Packages
+
+- Attributes in `packages.${system}` points to [derivation](https://nixos.org/manual/nix/stable/language/derivations.html) that can be used to build the package.
+- `nix build` will build the derivation of the package with the name `default`. Run `nix build .#<packageName>` to build a pacakge with a specific name.
+
+#### Apps
+
+- `apps.${system}.<appName>` refers to an attrset with two required attributes `type` and `program`. The `type` attribute determines how the program should be executed, For example, "shell" for a shell script, "python" for a Python script, or "app" for an executable. On the other hand, the `program` attribute is a string that represents the path in the Nix store where the executable is located.
+- `nix run` just takes the `executable` and based on `type` executes the `program` given by `apps.${system}.default`. Run `nix run .#<appName>` to run a specific app.
+
+#### DevShells
+
+- This puts you in a development shell with only the packages you need.
+- You can define how your shell will look using `pkgs.mkShell`.
+- `pkgs.mkShell` creates a derivation that is evaluated by the command `nix develop`
+- Derivation given by `devShells.${system}.default` is evaluated by default, you can also define custom `devShell` like `devShells.${system}.mydevShell` and run it using `nix develop .#mydevShell`
+
+It's time to now run the following commands and see the `flake.nix` in action:
+- `nix build`
+- `nix run`
+- `nix develop`
+
+## Nixify Haskell package
+
+We saw how to use `pkgs.hello` from `nixpkgs` in the basic example above, now it is time to define build instructions to build a package for the haskell project `todo-app`.\
+Tl;dr Here's the `flake.nix` for this section:
+```nix
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+  };
+  outputs = { self, nixpkgs }:
+    let
+      system = "aarch64-darwin";
+      pkgs = nixpkgs.legacyPackages.${system};
+      overlay = final: prev: {
+        todo-app = final.callCabal2nix "todo-app" ./. { };
+      };
+      myHaskellPackages = pkgs.haskellPackages.extend overlay;
+    in
+    {
+      packages.${system}.default = myHaskellPackages.todo-app;
+      apps.${system}.default = {
+        type = "app";
+        program = "${self.packages.${system}.default}/bin/todo-app";
+      };
+    };
+}
+```
+Let's break it down!
+### haskellPackages
+
+This is an attribute set that consists of all the haskell projects maintained in Nixpkgs. Since our local project will not be part of this, we have to add it. Technically you can just use  `packages.${system}.default = pkgs.${system}.haskellPackages.callCabal2nix "todo-app" ./. { };` but adding it to `haskellPackages` keeps it all in one place and consuming flakes will not have to think about where to find this package. 
```

Comment:
```suggestion
`pkgs.haskellPackages` is an attribute set consisting of all haskell packages maintained in nixpkgs. Since our local package (`todo-app`) is not already in it, we will have to add it manually. Technically you can just use  `packages.${system}.default = pkgs.${system}.haskellPackages.callCabal2nix "todo-app" ./. { };` but adding it to `haskellPackages` keeps it all in one place and any consuming flake will not have to think about where to find this package. 
```